File: November2022/Java/MaximumProfitInJobScheduling.java
// @saorav21994

class Solution {
    
    private class job {
        private int st;
        private int et;
        private int p;
        
        public job(int st, int et, int p) {
            this.st = st;
            this.et = et;
            this.p = p;
        }
        
    }
    
    public int jobScheduling(int[] startTime, int[] endTime, int[] profit) {
        List<job> list = new ArrayList<>();
        for (int i = 0; i < startTime.length; i++) {
            list.add(new job(startTime[i], endTime[i], profit[i]));
        }
        Collections.sort(list, (a,b)->(a.et-b.et));
        TreeMap<Integer, Integer> map = new TreeMap<>();
        int max = Integer.MIN_VALUE;
        for (job j : list) {
            Integer stime = map.floorKey(j.st);
            int curp = stime == null ? 0 : map.get(stime);
            max = Math.max(max, curp + j.p);
            map.put(j.et, max);
        }
        return max;
    }
}

File: November2022/C++/maximum-profit-in-job-scheduling.cpp
class Solution {
     int getNextIndex(vector<vector<int>>& array, int l, int currentJobEnd) {
         int n = array.size();
        int r = n-1;
        
        int result = n+1;
        
        while(l <= r) {
            int mid = l + (r-l)/2;
            
            if(array[mid][0] >= currentJobEnd) { //we can take this task
                result = mid;
                r = mid-1;
            } else {
                l = mid+1;
            }
        }
        
        return result;
    }
   
    int find(vector<int>&dp,vector<vector<int>>&time,int index){
        
        if(index>=time.size()){
            return 0;
        }
        
        if(dp[index]!= -1){
            return dp[index];
        }
        
        int take = 0;
        int notTake = 0;
        int in = getNextIndex(time,index+1,time[index][1]);

           take = time[index][2] + find(dp,time,in);    
       notTake= find(dp,time,index+1);
        
        return dp[index] = max(take,notTake);
    }
    
public:
    int jobScheduling(vector<int>& startTime, vector<int>& endTime, vector<int>& profit) {
 
        int n  = startTime.size();
        vector<vector<int>>time;
        for(int i=0;i<n;i++){
            time.push_back({startTime[i],endTime[i],profit[i]});
        }
        sort(time.begin(),time.end());
       vector<int>dp(n+1,-1);
 
         return    find(dp,time,0);
    
    }
};